File: backend/src/security/redaction.py
from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import Any


SECRET_KEYS = {
    "authorization",
    "gemini_api_key",
    "geminiapikey",
    "apikey",
    "api_key",
    "token",
    "x-api-key",
}


def is_secret_key(key: str) -> bool:
    normalized = key.replace("-", "_").lower()
    return normalized in SECRET_KEYS or normalized.endswith("_token")


def redact_value(value: Any) -> Any:
    if isinstance(value, Mapping):
        return {key: ("[REDACTED]" if is_secret_key(str(key)) else redact_value(item)) for key, item in value.items()}
    if isinstance(value, Sequence) and not isinstance(value, (str, bytes, bytearray)):
        return [redact_value(item) for item in value]
    return value


def redact_error_message(message: str) -> str:
    lowered = message.lower()
    if "key" in lowered or "token" in lowered or "authorization" in lowered:
        return "Secret-bearing request failed."
    return message

File: frontend/lib/redaction.ts
const REDACTED = "[REDACTED]"

const FIELD_VALUE_PATTERNS = [
  /("?(?:geminiApiKey|gemini_api_key|authorization|api[_ -]?key)"?\s*[:=]\s*)("[^"]*"|'[^']*'|[^\s,}\]]+)/gi,
  /((?:geminiApiKey|gemini_api_key|authorization|api[_ -]?key)\s+)([^\s,}\]]+)/gi,
]

export function redactSecrets(value: string, secrets: Array<string | undefined> = []) {
  let redacted = value

  for (const secret of secrets) {
    const normalized = secret?.trim()
    if (!normalized) {
      continue
    }
    redacted = redacted.split(normalized).join(REDACTED)
  }

  for (const pattern of FIELD_VALUE_PATTERNS) {
    redacted = redacted.replace(pattern, `$1${REDACTED}`)
  }

  return redacted
}

export function redactErrorMessage(error: unknown, secrets: Array<string | undefined> = []) {
  const message = error instanceof Error ? error.message : "Request failed."
  return redactSecrets(message, secrets)
}
